os/drivers/input/touch.c

```diff
@@ -0,0 +1,325 @@
+	DEBUGASSERT(buflen > 0);
+	DEBUGASSERT(filep);
+	inode = filep->f_inode;
+
+	DEBUGASSERT(inode && inode->i_private);
+	priv = inode->i_private;
+
+	/* Wait for semaphore to prevent concurrent reads */
+	touch_semtake(&priv->sem, false);
+
+	/* Read the touch data, only if screen has been touched or if we're waiting for touch up */
+	outlen = sizeof(struct touch_sample_s);
+	ret = priv->dev->touch_read(priv->dev, buffer);
+	touch_semgive(&priv->sem);
+	return ret < 0 ? ret : outlen;
+}
+
+static ssize_t touch_write(FAR struct file *filep, FAR const char *buffer, size_t buflen)
+{
+	//TODO
+	return OK;
+}
+
+#ifndef CONFIG_DISABLE_POLL
+void touch_pollnotify(FAR struct touch_dev_s *priv, pollevent_t eventset)
+{
+	int itr;
+	for (itr = 0; itr < CONFIG_TOUCH_NPOLLWAITERS; itr++) {
+		struct pollfd *fds = g_dev->fds[itr];
+		if (fds) {
+			fds->revents |= (fds->events & eventset);
+			if (fds->revents != 0) {
+				sem_post(fds->sem);
+			}
+		}
+	}
+}
+#endif
+
+void touch_notify(struct touch_dev_s *priv)
+{
+	touch_pollnotify(priv, POLLIN);
+}
+
+#ifndef CONFIG_DISABLE_POLL
+static int touch_poll(FAR struct file *filep, FAR struct pollfd *fds, bool setup)
+{
+	FAR struct touch_s *priv = filep->f_inode->i_private;
+	pollevent_t eventset;
+	int ret;
+	int i;
+	bool pending = false;
+	/* Some sanity checking */
+#ifdef CONFIG_DEBUG
+	if (!priv->dev || !fds) {
+		return -ENODEV;
+	}
+#endif
+	ret = touch_semtake(&priv->pollsem, true);
+	if (ret < 0) {
+		/* A signal received while waiting for access to the poll data
+		 * will abort the operation.
+		 */
+		return ret;
+	}
+
+	if (setup) {
+		/* This is a request to set up the poll. Find an available
+		 * slot for the poll structure reference
+		 */
+		for (i = 0; i < CONFIG_TOUCH_NPOLLWAITERS; i++) {
+			/* Find an available slot */
+			if (!priv->fds[i]) {
+				/* Bind the poll structure and this slot */
+				priv->fds[i] = fds;
+				fds->priv = &priv->fds[i];
+				fds->filep = (void *)filep;
+				break;
+			}
+		}
+
+		if (i >= CONFIG_TOUCH_NPOLLWAITERS) {
+			fds->priv = NULL;
+			fds->filep = NULL;
+			ret = -EBUSY;
+			goto errout;
+		}
+
+		eventset = 0;
+		touch_semtake(&priv->sem, false);
+		pending = priv->dev->is_touchSet(priv->dev);
+
+		if (pending) {
+			eventset |= (fds->events & POLLIN);
+		}
+		touch_semgive(&priv->sem);
+
+		if (eventset) {
+			touch_pollnotify((struct touch_dev_s *)&priv->dev, POLLIN);
+		}
+
+	} else if (fds->priv) {
+		/* This is a request to tear down the poll. */
+		struct pollfd **slot = (struct pollfd **)fds->priv;
+		/* Remove all memory of the poll setup */
+		*slot = NULL;
+		fds->priv = NULL;
+		fds->filep = NULL;
+	}
+errout:
+	touch_semgive(&priv->pollsem);
+	return ret;
+}
+#endif
+
+int touch_register(struct touch_dev_s *dev)
+{
+	g_dev = (struct touch_s *)kmm_zalloc(sizeof(struct touch_s));
+	if (!g_dev) {
+		touchdbg("ERROR: Memory allocation failed\n");
+	}
+	sem_init(&g_dev->sem, 0, 1);
+	sem_init(&g_dev->pollsem, 0, 1);
+	g_dev->dev = dev;
```

Comment:
How about following the user fish to the upper and lower parts like timer?

FAR struct timer_upperhalf_s *upper;

  upper->lower = lower;
  lower->upper = upper; or lower->parent = upper;

Reply:
upper is defined in drivers/*.c and lower is defined in tinyara/*.h, so how can lower use upper

Reply:
@ewoodev Is this enough? If yes, please mark it as resolved conversation.